File: android/src/main/java/com/rngooglemapsplus/MapHeatmapManager.kt
package com.rngooglemapsplus

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.TileOverlay
import com.google.maps.android.heatmaps.HeatmapTileProvider
import com.rngooglemapsplus.extensions.heatmapEquals
import com.rngooglemapsplus.extensions.heatmapNeedsRebuild

private class HeatmapState(
  var current: RNHeatmap,
) {
  var provider: HeatmapTileProvider? = null
  var overlay: TileOverlay? = null
}

class MapHeatmapManager(
  private val builder: MapHeatmapBuilder,
) {
  private var map: GoogleMap? = null
  private val states = mutableMapOf<String, HeatmapState>()
  private var destroyed = false

  fun attachMap(map: GoogleMap) =
    onUi {
      if (destroyed) return@onUi
      this.map = map
      states.values.filter { it.overlay == null }.forEach { addToMap(it) }
    }

  fun add(heatmap: RNHeatmap) =
    onUi {
      if (destroyed) return@onUi
      states.remove(heatmap.id)?.let { removeFromMap(it) }
      val state = HeatmapState(heatmap)
      states[heatmap.id] = state
      if (map != null) addToMap(state)
    }

  fun update(next: RNHeatmap) =
    onUi {
      if (destroyed) return@onUi
      val state = states[next.id] ?: return@onUi
      val prev = state.current
      if (prev.heatmapEquals(next)) return@onUi
      state.current = next
      val provider = state.provider ?: return@onUi
      val overlay = state.overlay ?: return@onUi
      if (prev.heatmapNeedsRebuild(next)) {
        removeFromMap(state)
        addToMap(state)
      } else {
        builder.update(prev, next, provider, overlay)
      }
    }

  fun remove(id: String) =
    onUi {
      states.remove(id)?.let { removeFromMap(it) }
    }

  fun destroy() =
    onUi {
      destroyed = true
      states.values.forEach { removeFromMap(it) }
      states.clear()
      map = null
    }

  private fun addToMap(state: HeatmapState) {
    val provider = builder.buildProvider(state.current)
    state.provider = provider
    state.overlay = map?.addTileOverlay(builder.build(state.current, provider))
  }

  private fun removeFromMap(state: HeatmapState) {
    state.overlay?.let {
      it.clearTileCache()
      it.remove()
    }
  }
}

File: ios/MapHeatmapManager.swift
import GoogleMaps
import GoogleMapsUtils

private final class HeatmapState {
  var current: RNHeatmap
  var overlay: GMUHeatmapTileLayer?

  init(current: RNHeatmap) {
    self.current = current
  }
}

final class MapHeatmapManager {
  private let builder: MapHeatmapBuilder
  private weak var mapView: GMSMapView?
  private var states: [String: HeatmapState] = [:]
  private var destroyed = false

  init(builder: MapHeatmapBuilder) {
    self.builder = builder
  }

  func attachMap(_ mapView: GMSMapView) {
    onMain {
      guard !self.destroyed else { return }
      self.mapView = mapView
      self.states.values.filter { $0.overlay == nil }.forEach { self.addToMap($0) }
    }
  }

  func add(_ heatmap: RNHeatmap) {
    onMain {
      guard !self.destroyed else { return }
      self.remove(id: heatmap.id)
      let state = HeatmapState(current: heatmap)
      self.states[heatmap.id] = state
      if self.mapView != nil {
        self.addToMap(state)
      }
    }
  }

  func update(_ next: RNHeatmap) {
    onMain {
      guard !self.destroyed else { return }
      guard let state = self.states[next.id] else { return }
      let prev = state.current
      if prev.heatmapEquals(next) { return }
      state.current = next
      guard let overlay = state.overlay else { return }
      if prev.heatmapNeedsRebuild(next) {
        self.removeFromMap(state)
        self.addToMap(state)
      } else {
        self.builder.update(prev, next, overlay)
      }
    }
  }

  func remove(id: String) {
    onMain {
      self.states.removeValue(forKey: id).map { self.removeFromMap($0) }
    }
  }

  func destroy() {
    onMain {
      self.destroyed = true
      self.states.values.forEach { self.removeFromMap($0) }
      self.states.removeAll()
      self.mapView = nil
    }
  }

  private func addToMap(_ state: HeatmapState) {
    guard let mapView = self.mapView else { return }
    let overlay = self.builder.build(state.current)
    overlay.map = mapView
    state.overlay = overlay
  }

  private func removeFromMap(_ state: HeatmapState) {
    state.overlay.map {
      $0.clearTileCache()
      $0.map = nil
    }
  }
}
